Improvements to gradients.

pyfisch opened the pull request:
This is a collection of commits improving the rendering of linear and radial gradients by making them conform more closely to the spec.

All commits are are independent and should work without the others.

These commits address the following issues:
* a956e3fd529715cc0ac39b23910f19e092c7c5a9 resolves #3908 but contains also some other necessary changes to `convert_gradient_stops`. The updated function has a few more copys but should be more correct. Maybe @pcwalton wants to comment since he has originally written the code.
* b230be8aaf318fb754cf58e5cd243087df2f7e0f partially solves #16638. (Partially because `border-image-outset` is not implemented. This is an older issue for border gradients: #15894.

To quickly catch regressions and see changes to gradients I have created [a set of twelve manual testcases](https://pyfisch.org/stuff/testcases-gradients.html) and placed them in a single file. Attached are two files. One shows how the gradients were rendered before the PR the other one with the changes applied.

![testcases-old](https://cloud.githubusercontent.com/assets/2781017/25580052/b433278e-2e7d-11e7-9396-500fef12eee0.png)
![testcases-new](https://cloud.githubusercontent.com/assets/2781017/25580051/b43222c6-2e7d-11e7-99ab-c0a2709baf41.png)

r? @emilio
and maybe also @jdm?

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16666)
<!-- Reviewable:end -->

pyfisch committed:
Implement -webkit-radial-gradient aliases.

In display_list_builder.rs handle cover and contain size keywords.

pyfisch committed:
Improve convert_gradient_stops function.

Implement Specification from CSS Images 3.

Improvements:

1. Stops with positions less than preceding stops are moved.
     A common pattern is blue `70%, white 0` for sharp transitions.
2. Stop runs are correct if first stop has a position.
    The list `black 0%, red, gold` now renders the same as `black, red, gold`.
    Other runs may also be corrected.
3. Absolute length are no longer capped to 100%.

highfive commented:
Heads up! This PR modifies the following files:
 * @emilio: components/layout/display_list_builder.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify layout code, but no tests are modified. Please consider adding a test!

pyfisch commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit fe7e7748b07b4866b38f5e97b35f8fde45826624 with merge 86cb19500c749f5b8cdfd832685772771221f9fe...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3763)

pyfisch committed:
Implement radial gradients for borders.

The property border-image-outset is not yet implemented.
Note: Also support repeating-linear-gradients for borders.

pyfisch committed:
Add fix_gradient_stops function.

Render gradients where the last stop is the same position
as the previous one like Chrome and Firefox do.

pyfisch committed:
Correct virtual gradient ray length of radial gradients.

The length should be the horizontal radius and not the
hypotenuse.

pyfisch commented:
r? @emilio 

@bors-servo try

The PR is finished now and awaiting review. :smile:

bors-servo commented:
:hourglass: Trying commit 4f17b17082a1a72c8f9470f22f8018f1cb0e41ab with merge 67da45ccca0152f710618eb75cd17659e7db36e5...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/6654), [arm32](http://build.servo.org/builders/arm32/builds/6869), [arm64](http://build.servo.org/builders/arm64/builds/6897), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7009), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3749), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3775), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7009), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/7041), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3698), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3765), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2197)
State: approved= try=True

pyfisch commented:
I have added three more testcases (number 13 to 15) to the example page.
* testcase 13 is a gradient with many layers rendered a bit wrong.
* testcase 14 highlights another bug in converting stops to absolute positions.
* testcase 15 shows that semi-transparent stops are rendered as opaque.

pyfisch committed:
Duplicate first gradient stop if necessary.

If the first stop of a non-repeating gradient is
not placed at offset 0.0 it is duplicated at this
position. This solves the problem of the first
stop being ignored if it is placed at the same
offset as the next stop.

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -608,57 +608,81 @@ fn convert_gradient_stops(gradient_items: &[GradientItem],
+    // Run the algorithm from
+    // https://drafts.csswg.org/css-images-3/#color-stop-syntax
+
+    // Step 1: If nothing else is specified the first stop is
```
I'd prefer to quote the spec directly, it makes it easier to grep:

```
// Step 1: If the first color stop does not have a position, set its position to 0%.
{
    // ...
}
// If the last color stop does not have a position, set its position to 100%.
{
    // ...
}
```

pyfisch replied:
Done for step 1.

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -608,57 +608,81 @@ fn convert_gradient_stops(gradient_items: &[GradientItem],
     // FIXME(#3908, pcwalton): Make sure later stops can't be behind earlier stops.
-    let stop_items = gradient_items.iter().filter_map(|item| {
+
+    // Only keep the color stops. Discard the color items.
```
"Only keep the color stop, discard the color interpolation hints."

pyfisch replied:
Thanks.

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -608,57 +608,81 @@ fn convert_gradient_stops(gradient_items: &[GradientItem],
     // Determine the position of each stop per CSS-IMAGES § 3.4.
     //
     // FIXME(#3908, pcwalton): Make sure later stops can't be behind earlier stops.
```
This FIXME can go away now, right?

pyfisch replied:
Obviously, I thought I had removed the FIXME.

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -608,57 +608,81 @@ fn convert_gradient_stops(gradient_items: &[GradientItem],
+
+    // Step 2: Move any stops placed before earlier stops to the
+    // same position as the preceding stop.
+    let mut state = stop_items.first().unwrap().position.unwrap();
```
instead of `state`, what about `last_stop_position`?

Also, we could move it to its own function to make this one a bit less dense, but as you prefer.

pyfisch replied:
Wouldn't `previous_stop_position` be a better name? `last_stop_position` can also be `stop_items.last()`.

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -608,57 +608,81 @@ fn convert_gradient_stops(gradient_items: &[GradientItem],
+    // Step 2: Move any stops placed before earlier stops to the
+    // same position as the preceding stop.
+    let mut state = stop_items.first().unwrap().position.unwrap();
+    for stop in stop_items.iter_mut() {
```
You can use `stop_items.iter_mut().skip(1)`, since there's no stop before the first one.

pyfisch replied:
Good catch.

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -608,57 +608,81 @@ fn convert_gradient_stops(gradient_items: &[GradientItem],
+    let mut state = stop_items.first().unwrap().position.unwrap();
+    for stop in stop_items.iter_mut() {
+        if let Some(pos) = stop.position {
+            if position_to_offset(state, length) > position_to_offset(pos, length) {
```
Perhaps renaming `length` to `available_length` would make this function a bit clearer, but no need to do it, I guess, since it's pre-existing.

pyfisch replied:
Inside  `fn position_to_offset` it is called `total_length`. I will go with this name.

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -673,6 +697,35 @@ fn convert_gradient_stops(gradient_items: &[GradientItem],
+/// stop is smaller than 100%. (Otherwise they ignore the last stop.)
+///
+/// Similarly the first stop is duplicated if it is not placed
+/// at the start of the virtual gradient ray.
```
Thanks for the comment, but... I guess the question is, what does the spec say? :)

pyfisch replied:
> If multiple color stops have the same position, they produce an infinitesimal transition from the one specified first in the rule to the one specified last. In effect, the color suddenly changes at that position rather than smoothly transitioning.

Well this is what the spec says about color stops with the same position. There are no special rules for the first and last stop. The way I understand the spec testcase 12 should have a green background with the circle in the middle like example 11 but bigger. Unfortunately  this is not what is implemented in Blink and Gecko. Maybe there is a note in the spec but I missed it completely. :confused: 

servo/webrender#1189 seems to do something similiar but on the webrender side. (I have not tested if it actually works like this function.

emilio replied:
Yeah, I remember to have seen that PR, thought it'd landed. We can land this for now and then remove/modify `fix_gradient_stops` if necessary. Thanks for the explanation!

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -1097,7 +1140,10 @@ impl FragmentDisplayListBuilding for Fragment {
 
-        let stops = convert_gradient_stops(stops, length, style);
+        let mut stops = convert_gradient_stops(stops, length, style);
+        if !repeating {
```
Can you elaborate on why not fixing them up if it's a repeating linear gradient, perhaps with a comment?

pyfisch replied:
Done for both linear and radial gradients.

emilio reviewed (commented):
Looks great overall, thanks a lot for doing this!

I've got a few nits, and one question.

pyfisch committed:
Address feedback by emilio.

pyfisch commented:
Updated.

emilio commented:
@bors-servo r+

Thanks a lot for this! It's a shame we can't land your test-cases as reftests btw :(.

bors-servo commented:
:pushpin: Commit f4fadc7 has been approved by `emilio`

<!-- @bors-servo r=emilio f4fadc78753dc2c7886b6ae1709431ee4488d058 -->

bors-servo commented:
:hourglass: Testing commit f4fadc78753dc2c7886b6ae1709431ee4488d058 with merge 3f1ae6425588bbac37da2e0f8aa2a708f51fa988...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/6762), [arm32](http://build.servo.org/builders/arm32/builds/6978), [arm64](http://build.servo.org/builders/arm64/builds/7010), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7113), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3859), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3878), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7109), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/7146), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3802), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3872), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2313)
Approved by: emilio
Pushing 3f1ae6425588bbac37da2e0f8aa2a708f51fa988 to master...
